man/man1/lpjcheck.1
.TH lpjcheck 1  "USER COMMANDS"
.SH NAME
lpjcheck \- Checks syntax of LPJmL configuration files
.SH SYNOPSIS
.B lpjcheck
[\-h] [\-v] [\-q] [\-nocheck] [-vv] [\-ofiles] [\-param] [-nopp] [-pp cmd] [\-pedantic] [\-couple host[:port]] [\-outpath \fIdir\fP]
[\-inpath \fIdir\fP] [\-restartpath \fIdir\fP] [[\-Dmacro[=value]] [\-I\fIdir\fP] ...]
\fIfilename\fP
.SH DESCRIPTION
Program checks the syntax of LPJmL configuration files. It is checked additionally whether all input files and output directories are present.
\fBlpjcheck\fP will print the total amount of disk space used for the output files.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-v,\--version
Print LPJmL version.
.TP
\-q,\--quiet
Only error messages are displayed, other output is suppressed.
.TP
\-nocheck
Do not check for input/output files.
.TP
\-vv
Verbosely print the actual values during reading of the configuration files.
.TP
\-ofiles
List only all available output variables.
.TP
\-param
Print LPJmL parameter.
.TP
\-nopp
Disabling preprocessing of configuration file by \fBcpp\fP.
.TP
\-pp cmd
Set preprocessor program to cmd. Default is \fBcpp\fP.
.TP
\-pedantic
Return error on warnings.
.TP
\-couple host[:port]
Set host where the coupled model is running. Default is localhost. Port is the port number for ingoing and outgoing data. Default port number is 2224.
.TP
\-outpath \fIdir\fP
Set the output directory path. The path is added to the output filenames if they do not contain an absolute path.
.TP
\-inpath \fIdir\fP
Set the input directory path. The path is added to the input filenames if they do not contain an absolute path.
.TP
\-restartpath \fIdir\fP
Set the restart directory path. The path is added to the restart filenames if they do not contain an absolute path.
.TP
\-Dmacro[=value]
Define macro for the preprocessor of the configuration file.
.TP
\-I\fIdir\fP
Define include directory for the preprocessor of the configuration file.
.TP
.I filename
Name of configuration file.
.SH EXAMPLES
.TP
Check configuration file lpjml.cjson with input files located in the /p/projects/lpjml file system
.B lpjcheck
\-inpath /p/projects/lpjml lpjml.cjson
.PP
.SH ENVIRONMENT
.TP
LPJNOPP
Disables preprocessor if set to "true", same as -nopp option.
.TP
LPJPEDANTIC
Enables pedantic mode if set to "true", same as -pedantic option.
LPJPREP
Defines preprocessor command for LPJmL configuration file, default is \fBcpp\fP. Same as '-pp' option.
.TP
LPJROOT
Defines the root directory for LPJmL. This directory is added to the
include directories of the preprocessor
.TP
LPJINPATH
Path appended to the input filenames. Only done for filenames without absolute path.
.TP
LPJOPTIONS
Runtime options for the preprocessor of LPJmL configuration files
.TP
LPJOUTPATH
Path appended to the output filenames. Only done for filenames without absolute path.
.TP
LPJRESTARTPATH
Path appended to the restart filenames. Only done for filenames without absolute path.

.SH EXIT STATUS
.B
lpjcheck
returns a zero exit status if the configuration file has no syntax errors and all input files and output directories exist.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), lpjfiles(1), lpjsubmit(1)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Stephen Wirth (wirth@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jannes Breier (breier@pik-potsdam.de)

### Added

- Added benchmark folder with with two R scripts (`simulate_default.R`, `benchmark_default.R`) to simulate default scenarios and run benchmark.

### Changed

- Argument `-couple $(hostname)` added to call of lpjml in `slurm.jcf` if `-couple` option is set for `lpjsubmit`.
- Wait for coupled program to finish added in slurm script.
- Fixed several potential floating point exceptions in `nuptake_tree.c`, `nuptake_grass.c`, `nuptake_crop.c` and `littersom.c`
- Printing of balance error messages replaced by call to `fail()` function in `allocation_grass.c`, `litter_update_grass.c`, `turnover_grass.c` and `update_annual_cell.c`.

### Removed

- Support for cloudiness and short wave only setting for climate input removed, only `"radiation"` and `"radiation_lwdown"` allowed.
- Output `"sun"` and `"sun_image"` removed.
- Unused files `nooutput_gbw.c` and `update_monthly.c` removed.
- Unused array `layer_exists` in soil struct removed.

### Fixed

- Function `openconfig.c` corrected to avoid segmentation violation if lpjml/lpjcheck is called with `-couple hostname` option.
- Changed `nupsum==0` to `nupsum<epsilon` in `nuptake_crop.c`, `nuptake_grass.c` and `nuptake_tree.c` to avoid potential FPEs.
- Function `cutpfts.c` corrected to close carbon balance.
- Water balance check in `infil_perc.c` correcte to avoid false water balance errors.
- Balance error messages corrected in `cultivate.c`.


## [6.0.2] - 2026-01-17

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jannes Breier (breier@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- Warning message added in `lpjml`, `lpjcheck`,`cdf2soil`, `txt2grid` and `cdf2coord` if cell size cannot be resolved by grid data in short format.
- Function `isfloatcoord()` added in `coord.c` to check whether coordinate must be stored as a float variable, test function `test_isfloatcoord.c` added.
- Test functions for `vmaxlimit_tree`, `vmaxlimit_grass`, `vmaxlimit_crop`, `bisect`, `leftmostzero` and `getbufferavg` functions added.

### Changed

- Error about first simulation year is less than first year of climate data replaced by warning in `initclimate.c` and `filesexist.c`.
- First and last simulation year in spinup changed from 1901 to 1700 in `lpjml_config.cjson`.
- Number of spinup years changed from 4000 to 3799 in `lpjml_config.cjson` in order to get bit-identical results.
- Test files for unit test and shell scripts added to tar and zip file in `Makefile`.
- First year of climate data set instead of first simulation year used in `initsoiltemp.c`.
- Unused argument `ncft` removed from `check_stand_fracs.c` and `update_wetland.c`.
- Redundant calculation of stand fraction sum removed in `check_stand_fracs.c`.

### Removed

- Unused source file `src/lpj/part.c` removed.

### Fixed

- Expected number of bands set correctly for raw input files in `openinputdata.c`.
- Hard-coded filename `test.lpj` replaced by temporary filename in `test_append.c`, `test_getnoread.c`, `test_indexarray.c`, `test_isdefined.c`, `test_isnull.c`, `test_restart.c`, and `test_writename.c` in order to allow parallel execution of tests.
- Arguments corrected in calls of lpjml functions in `lpj_poem.c`.
- Missing `-print_noread` option added in `INSTALL`.
- `setupannual_grid.c` corrected to compile with `-DIMAGE -DCOUPLED` flag.
- Error message for different datatypes corrected in `lpjcat.c`.
- Variables initialized to zero and NULL in `phenology_gsi.c`, `lpjcat.c` and `restart2yaml.c` to avoid warnings in `gcc`.
- Man page for `lpjml`, `lpjcheck`, `lpjrun`, `lpjfiles`, `json2restart` and `restart2yaml` corrected.
- Missing option `-nooutput` added to man page of `lpjfiles`.
- Help text for `-h` option of `lpjfiles`, `restart2yaml`, `printclm` and `regridclm` corrected.
- Check for NULL pointer for `pfts` array corrected in `lpjcat.c`.
- Function `fread()` replaced by `freadint()` in `regridirrig.c` to account for different byte order.
- Missing deallocation of `items` array added in `sendhash.c`.
- Filename for population density input corrected in `regridlpj`.
- Header files added for unit tests to compile tests successfully.
- Function `albedo_crop.c` and `fwritecell.c` changed to compile with `-DCOUPLING_WITH_FMS` setting.
- Misplaced check for NULL pointer for wind removed in `dailyclimate.c`, unnecessary check for NULL pointer for wind removed.
- Missing `delta_prec` and `delta_temp` filenames added in `fprintfiles.c`.
- Missing dependency on `climate.h` added to Makefile in `src/landuse`.
- Double occurrence of `"outputyear"` removed in `lpjml_config_pnv.cjson`.


## [6.0.1] - 2025-12-18

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Changed

- Declaration of `mixsoilenergy()` moved from `setaside.c` to `landuse.h`.

### Removed

- Unused functions `snow.old()`, `soiltemp()` and related functions removed.
- Setting without permafrost removed
- Soil variables `state[]`, `alag`, `amp`, and `meanw1` removed.


## [6.0.0] - 2025-11-25


src/utils/lpjcheck.c
/**************************************************************************************/
/**                                                                                \n**/
/**                 l  p  j  c  h  e  c  k  .  c                                   \n**/
/**                                                                                \n**/
/**     Program checks syntax of LPJ configuration files and tests on              \n**/
/**     existence of input files and output directories.                           \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"
#include "grass.h"
#include "tree.h"
#include "crop.h"

#define NTYPES 3 /* number of PFT types: grass, tree, crop */
#define USAGE "\nUsage: %s [-h] [-v] [-q] [-nocheck] [-ofiles] [-param] [-vv]\n"\
              "       [-couple hostname[:port]] [-pedantic]\n"\
              "       [-outpath dir] [-inpath dir] [-restartpath dir]\n"\
              "       [-nopp] [-pp cmd] [[-Dmacro[=value]] [-Idir] ...] filename\n"
#define LPJ_USAGE USAGE  "\nTry \"%s --help\" for more information.\n"

int main(int argc,char **argv)
{
  /* Create array of functions, uses the typedef of Pfttype in config.h */
  Pfttype scanfcn[NTYPES]=
  {
    {name_grass,fscanpft_grass},
    {name_tree,fscanpft_tree},
    {name_crop,fscanpft_crop}
  };
  Config config;         /* LPJ configuration */
  int rc=0;              /* return code of program */
  Bool isout,check;
  const char *progname;
  const char *title[4];
  String line,line2;
  FILE *file;
  initconfig(&config);
  isout=check=TRUE;
  progname=strippath(argv[0]);
  if(argc>1)
  {
    if(!strcmp(argv[1],"-q") || !strcmp(argv[1],"--quiet")) /* checks for '-q' flag */
    {
      argc--; /* adjust command line options */
      argv++;
      isout=FALSE; /* no output */
    }
    else if(!strcmp(argv[1],"-v") || !strcmp(argv[1],"--version"))
    {
      puts(getversion());
      return EXIT_SUCCESS;
    }
    else if(!strcmp(argv[1],"-h") || !strcmp(argv[1],"--help"))
    {
      file=popen("more","w");
      if(file==NULL)
        file=stdout;
      fprintf(file,"     ");
      rc=fprintf(file,"%s (" __DATE__ ") Help",
              progname);
      fprintf(file,"\n     ");
      frepeatch(file,'=',rc);
      fprintf(file,"\n\nCheck syntax of LPJmL version %s configuration (*.cjson) files\n",getversion());
      fprintf(file,USAGE,progname);
      fprintf(file,"\nArguments:\n"
             "-h,--help           print this help text\n"
             "-v,--version        print LPJmL version\n"
             "-q,--quiet          print error messsages only\n"
             "-vv                 verbosely print the actual values during reading of the\n"
             "                    configuration files\n"
             "-pedantic           stops on warnings\n"
             "-ofiles             list only all available output variables\n"
             "-param              print LPJ parameter\n"
             "-nopp               disable preprocessing\n"
             "-pp cmd             set preprocessor program. Default is '" cpp_cmd "'\n"
             "-couple host[:port] set host and port where coupled model is running\n"
             "-outpath dir        directory appended to output filenames\n"
             "-inpath dir         directory appended to input filenames\n"
             "-restartpath dir    directory appended to restart filename\n"
             "-Dmacro[=value]     define macro for preprocessor of configuration file\n"
             "-Idir               directory to search for include files\n"
             "filename            configuration filename\n\n"
             "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n");
      if(file!=stdout)
        pclose(file);
      return EXIT_SUCCESS;
    }
  }
  if(argc>1 && !strcmp(argv[1],"-nocheck"))
  {
    argc--; /* adjust command line options */
    argv++;
    check=FALSE; /* no checking */
  }
  if(isout)
  {
    snprintf(line,78-10,
             "%s (" __DATE__ ")",progname);
    snprintf(line2,78-10,"Checking configuration file for LPJmL version %s",getversion());
    title[0]=line;
    title[1]=line2;
    title[2]="(C) Potsdam Institute for Climate Impact Research (PIK),";
    title[3]="see COPYRIGHT file";
    banner(title,4,78);
  }

  if(readconfig(&config,scanfcn,NTYPES,NOUT,&argc,&argv,LPJ_USAGE))
  {
    fail(READ_CONFIG_ERR,TRUE,FALSE,"Cannot process configuration file");
  }
  else
  {
    if(argc)
      fputs("WARNING018: Arguments listed after configuration filename, will be ignored.\n",stderr);
    if(config.ofiles)
    {
      fprintoutputvar(stdout,config.outnames,NOUT,config.npft[GRASS]+config.npft[TREE],config.npft[CROP],&config);
      return EXIT_SUCCESS;
    }

    if(isout)
    {
      /* print LPJ configuration on stdout if '-q' option is not set */
      printconfig(config.npft[GRASS]+config.npft[TREE],
                  config.npft[CROP],&config);
    }
    if(config.nall>0 && config.n_out)
    {
      config.nall=config.total=config.ngridcell;
      printf("Estimated disk usage for output: ");
      printintf((int)(outputfilesize(&config)/(1024*1024)));
      printf(" MByte\n");
    }
    if(check)
      rc=(filesexist(&config,isout)) ? EXIT_FAILURE : EXIT_SUCCESS;
    else
      rc=EXIT_SUCCESS;
  }
  freeconfig(&config);
  return rc;
} /* of 'main' */
